### ISSUE TITLE
SQLite: Reloading article after having added a selector does not work

### ISSUE THREAD

**Issue by Massedil** [FreshRSS/FreshRSS#3130]
**How to reproduce :**

- Add a new feed
- Edit this feed
- Add a selector
- Check that the selector gives the good content with the new "eye" button
- Click on Validate ("Valider", in French)

Then go again to edit the feed :
- Click on Reload articles ("Recharger les articles", in French)

**Actual :**

The reloading is ***very slow***  but the feed does not have the website content expected.

**Expected :**

All the feeds contain the content from the original website

**Note :**

For the new articles published after, the website content will be well retrieved.

**Comment by aledeg** [comment]
Consider sharing more information like the feed, the CSS selector.
You can use the preview action before reloading all articles to validate your query.

Remember to clear the cache before reloading articles.

**Comment by Massedil** [comment]
I just retried now and I still can reproduce.

1. Add this feed into FreshRSS : `https://getsession.org/feed/`.
2. Now edit the feed and add a CSS selector : `#main .post`. 
3. Preview action works fine : I can see the whole article.
4. Click on "Validate"
5. Then, as you suggested, click on "clear cache" (it does nothing visible and does not solve the problem)
6. Click on "Reload all articles" :
    - I have to wait between 30 seconds and 1 minute (why is it so long ?) and then it successfully loads the feed panel 
    - But the articles are not updated with the full article contents

Version: 1.17.1-dev

**Comment by aledeg** [comment]
So, I've tried what you've done and there is a problem.
But when I check the logs, I have the following error: 
```
SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=1938 in feed 43
```
Do you have such thing?

**Comment by aledeg** [comment]
I've made some more tests and the content is properly retrieved by the CSS selector. The problem arise when saving the content in the database.

**Comment by Massedil** [comment]
> Do you have such thing?

Yes, I also have this : 

```
[Wed, 17 Feb 2021 19:54:02 +0100] [error] --- SQL error updateEntry: HY000: 5 database is locked while updating entry with GUID https://getsession.org/?p=2109 in feed 651
[Wed, 17 Feb 2021 19:54:02 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=2103 in feed 651
[Wed, 17 Feb 2021 19:54:03 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=2083 in feed 651

[...30 identical lines deleted...]

[Wed, 17 Feb 2021 19:54:28 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=964 in feed 651
[Wed, 17 Feb 2021 19:55:45 +0100] [error] --- SQL error updateEntry: HY000: 5 database is locked while updating entry with GUID https://getsession.org/?p=2109 in feed 651
[Wed, 17 Feb 2021 19:55:46 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=2103 in feed 651

[...30 identical lines deleted...]

[Wed, 17 Feb 2021 19:56:11 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=967 in feed 651
[Wed, 17 Feb 2021 19:56:12 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=964 in feed 651
[Wed, 17 Feb 2021 20:01:52 +0100] [error] --- SQL error updateEntry: HY000: 5 database is locked while updating entry with GUID https://getsession.org/?p=2109 in feed 651
[Wed, 17 Feb 2021 20:01:52 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=2103 in feed 651
[Wed, 17 Feb 2021 20:01:53 +0100] [error] --- SQL error updateLastUpdate: database is locked
[Wed, 17 Feb 2021 20:01:53 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=2083 in feed 651

[...30 identical lines deleted...]

[Wed, 17 Feb 2021 20:02:18 +0100] [error] --- SQL error updateEntry: HY000: 21 bad parameter or other API misuse while updating entry with GUID https://getsession.org/?p=964 in feed 651

```

**Comment by aledeg** [comment]
Good! Now it's just a matter of finding what is causing the issue.

**Comment by aledeg** [comment]
I narrowed it down a bit further. When I delete the article from the database and I update the feed, the full content is retrieved and properly stored in the database.

**Comment by aledeg** [comment]
The previous test was made by modifying data directly in the database.
I've done another test where I deleted every article but from the interface then I've updated the feed. Now, I have all articles with their full content.

I am aware that you might not be able to do that if you want to keep the history. But it works.
We need to find out why this is not working while reloading :/

**Comment by Alkarex** [comment]
I have not looked into it, but some errors might be due to having already opened a transaction, or a bad prepared statement. Which PHP version(s) are you using more precisely, and database?

**Comment by aledeg** [comment]
For my tests, I am using Sqlite and PHP8.0.1 from the official docker container

**Comment by Massedil** [comment]
> I am aware that you might not be able to do that if you want to keep the history. But it works.

Thanks, I just wanted to signal the bug. Nothing blocking for me.

> Which PHP version(s) are you using more precisely, and database?

PHP 7.3 and SQLite from Debian buster for me.

**Comment by aledeg** [comment]
So when the following line is commented, there is no more errors:
https://github.com/FreshRSS/FreshRSS/blob/0e6ad01dbfc396bee63f2e100df75bcd6488ed16/app/Controllers/feedController.php#L744

Going deeper, in the `updateEntry` method, if I comment the following line or if I force its content, I have errors:
https://github.com/FreshRSS/FreshRSS/blob/0e6ad01dbfc396bee63f2e100df75bcd6488ed16/app/Models/EntryDAO.php#L208

So my conclusion for today is that it's not related to the retrieved content but to the update method itself.

To be continued...

**Comment by Alkarex** [comment]
I do not have time to look more at it now, but here is a draft to help debugging https://github.com/FreshRSS/FreshRSS/pull/3461

**Comment by aledeg** [comment]
Using the same connection works! There must be some locking mechanism involved between the call that prevent the update.

**Comment by Alkarex** [comment]
I think this (having two connections in parallel) has been tested with SQLite, but I cannot remember for sure. It would be interesting to test (again) with PostgreSQL, MySQL, and SQLite.

**Comment by Alkarex** [comment]
The problem is limited to SQLite: works fine with PostgreSQL and MySQL.

**Comment by Alkarex** [comment]
Fixed in https://github.com/FreshRSS/FreshRSS/pull/3461 , tests welcome

**Comment by Massedil** [comment]
> Fixed in #3461 , tests welcome

Just updated to latest dev version, and I can confirm it solves the problem. Thanks !
